sdk/log/batch.go

```diff
@@ -43,39 +59,138 @@ func NewBatchingProcessor(exporter Exporter, opts ...BatchingOption) *BatchingPr
 		exporter = defaultNoopExporter
 	}
 	cfg := newBatchingConfig(opts)
+
+	// TODO: Add polling to the BatchingProcessor.
+
+	ctx, cancel := context.WithCancel(context.Background())
 	return &BatchingProcessor{
 		exporter: exporter,
 
 		maxQueueSize:       cfg.maxQSize.Value,
 		exportInterval:     cfg.expInterval.Value,
 		exportTimeout:      cfg.expTimeout.Value,
 		exportMaxBatchSize: cfg.expMaxBatchSize.Value,
+
+		batch:  newBatch(cfg.maxQSize.Value),
+		ctx:    ctx,
+		cancel: cancel,
 	}
 }
 
+// enqueueFunc is used for testing until #5105 is merged and integrated.
+var enqueueFunc = func(_ context.Context, _ []Record, ch chan error) {
+	if ch != nil {
+		ch <- nil
+	}
+}
+
+// enqueue attempts to enqueue an export. If the exportCh is full, the export
+// will be dropped and an error logged.
+func (b *BatchingProcessor) enqueue(ctx context.Context, r []Record, rCh chan error) {
+	// TODO (#5105): Implement this. Until this is implemented call enqueueFunc
+	// so the rest of the BatchingProcessor can be tested.
+	enqueueFunc(ctx, r, rCh)
+}
+
 // OnEmit batches provided log record.
-func (b *BatchingProcessor) OnEmit(ctx context.Context, r Record) error {
-	// TODO (#5063): Implement.
+func (b *BatchingProcessor) OnEmit(_ context.Context, r Record) error {
+	if b.stopped.Load() {
+		return nil
+	}
+	if flushed := b.batch.Append(r); flushed != nil {
+		b.enqueue(b.ctx, flushed, nil)
+	}
 	return nil
 }
 
-// Enabled returns true.
+// Enabled returns if b is enabled.
 func (b *BatchingProcessor) Enabled(context.Context, Record) bool {
-	return true
+	return !b.stopped.Load()
 }
 
 // Shutdown flushes queued log records and shuts down the decorated exporter.
 func (b *BatchingProcessor) Shutdown(ctx context.Context) error {
-	// TODO (#5063): Implement.
-	return nil
+	if b.stopped.Swap(true) {
+		return nil
+	}
+
+	resp := make(chan error, 1)
+	b.enqueue(ctx, b.batch.Flush(), resp)
+
+	// Cancel all exports and polling.
+	b.cancel()
```

Comment:
I am not sure why we should cancel pending exports. I think we should still give a chance to have them completed until export timeout is not reached.

Reply:
This cancels the context for the polling. This is not the context used in the last flush.

Reply:
```suggestion
	// Cancel all subsequent exports and polling.
	b.cancel()
```